### ISSUE TITLE
`core::0192::no-markdown-headings` false positive when appearing in a code block

### ISSUE THREAD

**Issue by Paulchen5** [googleapis/api-linter#1587]
I am linting my `.proto` files and got the following message:

```yaml
    - message: Comments should not include Markdown headings.
      location:
        start_position:
            line_number: 398
            column_number: 3
        end_position:
            line_number: 398
            column_number: 22
        path: ./google/api/client.proto
      rule_id: core::0192::no-markdown-headings
      rule_doc_uri: https://linter.aip.dev/192/no-markdown-headings
```
on this code:

```proto
  // The fully qualified name of the method, for which the options below apply.
  // This is used to find the method to apply the options.
  //
  // Example:
  //
  //    publishing:
  //      method_settings:
  //      - selector: google.storage.control.v2.StorageControl.CreateFolder
  //        # method settings for CreateFolder...
  string selector = 1;
```

The code comes from `./google/api/client.proto`

**Comment by CuriousStork** [comment]
Hello, my three cents here, this is also reproduced when you wrap the `#` character into markdown fenced code block, like,

```protobuf
  // The fully qualified name of the method, for which the options below apply.
  // This is used to find the method to apply the options.
  //
  // Example:
  //
  // ```yaml
  //    publishing:
  //      method_settings:
  //      - selector: google.storage.control.v2.StorageControl.CreateFolder
  //        # method settings for CreateFolder...
  // ```
  string selector = 1;
```

**Comment by noahdietz** [comment]
Hi @Paulchen5 thanks for the issue.

One clarifying question: Are you supplying `./google/api/client.proto` as input to be linted? Or only as an dependency to resolve impots?

> Hello, my three cents here, this is also reproduced when you wrap the # character into markdown fenced code block, like,

Yeah the code for this is rather naive - it just regexes for `#`, doesn't use a markdown parser. I'm really hesitant to make API Linter a markdown parser as well...I've pushed back on adding other such rules b.c I really think it should be done by a dedicated tool (even though AIP-192 has guidance).

Not sure what the best course of action is just yet...

**Comment by paulchen5** [comment]
Hmmm unfortunately I can't reproduce the issue. I may have changed my setup however now there is no waring about this anymore.

Probably before the change it was supplied by the input. However I think I was just confused that linting Googles own stuff with their own linter resulted in some problems😅

**Comment by noahdietz** [comment]
>  However I think I was just confused that linting Googles own stuff with their own linter resulted in some problems😅

You will get many findings linting the contents of googleapis for #reasons 😄 

Ok I'm going to keep this open but change the title and type to reflect the behavior @CuriousStork mentioned in https://github.com/googleapis/api-linter/issues/1587#issuecomment-3689166622

**Comment by noahdietz** [comment]
Off the top of my head I think headings in comments would appear as something like:
```proto
// # FooService
// This is the FooService
//
// ## Usage
// This is how you use the FooService
//
// ## Pricing
// The FooService costs a lot of money
service FooService {
```

i.e. a single leading space line rather than deeply indented line like the current regex matches on  https://github.com/googleapis/api-linter/blob/cd1ae06d3b4c995d5c8b8d15624c2f391af06654/rules/aip0192/no_markdown_headings.go#L43

I'm thinking we update the regex to simply check for something like `^[#]+ ` (the single space following the `//` should be stripped by comment conversion to simple text).

This allows indented code blocks to be correctly skipped.

**Comment by noahdietz** [comment]
Taking a stab in https://github.com/googleapis/api-linter/pull/1593

**Comment by CuriousStork** [comment]
Hello @noahdietz 

> I'm really hesitant to make API Linter a markdown parser as well

I completely share this concern. However, philosophically speaking, AIPs inherently imply a knowledge of Markdown. For example, https://google.aip.dev/192#formatting

> Any formatting in comments must be in [CommonMark](https://commonmark.org/). Headings and tables must not be used

In an ideal world, if the linter were to cover all AIP checks, it would naturally have to understand Markdown as well.

Of course, we don’t live in an ideal world, so the linter doesn’t necessarily have to support everything (at least not yet) :)
